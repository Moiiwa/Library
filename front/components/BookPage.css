.book{
    display: flex;
    justify-content: center;
}

.book-info{
    padding: 2rem;
}

.book-cover{
    width: 300px;
    height: 500px;
    padding: 2rem;
}

.description{
    width: 600px;
}