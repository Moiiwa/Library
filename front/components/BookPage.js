import React from 'react'
import './BookPage.css'

class BookPage extends React.Component {

    constructor(props) {
        super(props);
        this.book = {
            bookCover: 'https://images-na.ssl-images-amazon.com/images/I/61ZKNw0xixL.jpg',
            bookTitle: 'Book cover design',
            authors: ["John", "Dalb"],
            description: "Lorem Ipsum - это текст-\"рыба\", часто используемый в печати и вэб-дизайне. Lorem Ipsum является стандартной \"рыбой\" для текстов на латинице с начала XVI века. В то время некий безымянный печатник создал большую коллекцию размеров и форм шрифтов, используя Lorem Ipsum для распечатки образцов. Lorem Ipsum не только успешно пережил без заметных изменений пять веков, но и перешагнул в электронный дизайн. Его популяризации в новое время послужили публикация листов Letraset с образцами Lorem Ipsum в 60-х годах и, в более недавнее время, программы электронной вёрстки типа Aldus PageMaker, в шаблонах которых используется Lorem Ipsum."
        }
    }

    render() {
        return (
            <div className="book">

                <img className={"book-cover"} src={this.book.bookCover}  alt={"Book Cover"}/>

                <div className="{book-info}">
                    <h1>{this.book.bookTitle}</h1>
                    <b>Authors: </b>
                    <p>{this.book.authors}</p>
                    <b>Description:</b>
                    <p className={"description"}>{this.book.description}</p>
                </div>
            </div>
        )
    }

}

export default BookPage;