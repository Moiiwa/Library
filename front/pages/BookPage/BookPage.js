import React from 'react'
import { getBook, updateSharingStatus, updateSellingStatus } from '../../api/book'
import InfoField from '../../components/InfoField/InfoField'

import './BookPage.css'
class BookPage extends React.Component {

    constructor(props) {
        super(props);
        this.noCover = "https://lightning.od-cdn.com/static/img/no-cover_en_US.a8920a302274ea37cfaecb7cf318890e.jpg"
        this.state = {
            id: this.props.match.params.id,
            loading: true,
            bookData: null
        };
    }

    getBook = async () => {
        const id = this.state.id
        await getBook(id)
            .then(data => {
                this.setState({
                    loading: false,
                    bookData: data
                });
            })
    }

    componentDidMount() {
        this.getBook();
    }

    changeSellingStatus = async () => {
        let data = this.state.bookData
        let oppStatus = !data.sellingStatus
        await updateSellingStatus(oppStatus, this.state.id)
        data.sellingStatus = oppStatus
        this.setState({
            bookData: data
        });
    }

    changeSharingStatus = async () => {
        let data = this.state.bookData
        let oppStatus = !data.rentingStatus
        await updateSharingStatus(oppStatus, this.state.id)
        data.rentingStatus = oppStatus
        this.setState({
            bookData: data
        });
    }

    render() {
        return (
            <div className="book">
                <img className={"book-cover"} src={this.noCover} alt={"Book Cover"} />
                {this.state.loading ?
                    <div>Loading...</div> :
                    <div>
                        <div className="{book-info}">
                            <h1>{this.state.bookData.title}</h1>
                        </div>
                        {this.state.bookData.author !== undefined ?
                            <InfoField label={'Authors:'} value={this.state.bookData.author} />
                            :
                            <InfoField label={'Authors:'} value={'no'} />}
                        {this.state.bookData.publisher !== undefined ?
                            <InfoField label={'Publisher:'} value={this.state.bookData.publisher} />
                            : null}
                        {this.state.bookData.publishedDate !== undefined ?
                            <InfoField label={'Published date:'} value={this.state.bookData.publishedDate} />
                            : null}
                        {this.state.bookData.description !== undefined ?
                            <InfoField label={'Description:'} value={this.state.bookData.description} />
                            : <InfoField label={'Description:'} value={'no'} />}
                        {this.state.bookData.number_of_pages !== undefined ?
                            <InfoField label={'Pages:'} value={this.state.bookData.number_of_pages} />
                            : null}
                        <div>
                            <input type="checkbox" id='sell' checked={this.state.bookData.sellingStatus} onChange={this.changeSellingStatus} />
                            <label for='sell'>Allow selling</label>
                        </div>
                        <div>
                            <input type="checkbox" id='share' checked={this.state.bookData.rentingStatus} onChange={this.changeSharingStatus} />
                            <label for='share'>Allow sharing</label>
                        </div>
                    </div>
                }
            </div>
        )
    }

}

export default BookPage;