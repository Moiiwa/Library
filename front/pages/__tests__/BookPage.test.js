import React from 'react';
import BookPage from '../BookPage/BookPage';
import toJson from 'enzyme-to-json';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('BookPage', () => {

    const props = {
        match: {
            params: {
                id: 1
            }
        }
    }

    let mockSellingStatus;
    let mockRentingStatus;

    beforeEach(() => {
        mockSellingStatus = jest.fn();
        mockRentingStatus = jest.fn()
    });

    it('componentDidMount() works', () => {
        const wrapper = shallow(<BookPage {...props} />);
        const instance = wrapper.instance();
        jest.spyOn(instance, 'getBook');
        instance.componentDidMount();
        expect(instance.getBook).toHaveBeenCalledTimes(1);
    });

    it('renders correctly', () => {
        const wrapper = shallow(<BookPage {...props} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('update selling status', () => {
        const wrapper = shallow(<BookPage {...props} />);
        wrapper.setState({
            loading: false,
            bookData: {
                title: 'title',
                sellingStatus: true
            }
        });
        const instance = wrapper.instance()
        const changeSellingStatusMock = jest.spyOn(instance, 'changeSellingStatus')
        instance.forceUpdate();

        const checkbox = wrapper.find('#sell')
        checkbox.simulate('change', { target: { checked: true } });
        expect(changeSellingStatusMock).toHaveBeenCalled()
    });

    it('update renting status', () => {
        const wrapper = shallow(<BookPage {...props} />);
        wrapper.setState({
            loading: false,
            bookData: {
                title: 'title',
                rentingStatus: true
            }
        });
        const instance = wrapper.instance()
        const changeSharingStatusMock = jest.spyOn(instance, 'changeSharingStatus')
        instance.forceUpdate();

        const checkbox = wrapper.find('#share')
        checkbox.simulate('change', { target: { checked: true } });
        expect(changeSharingStatusMock).toHaveBeenCalled()
    });
});