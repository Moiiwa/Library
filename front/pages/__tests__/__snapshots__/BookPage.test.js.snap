// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookPage renders correctly 1`] = `
<div
  className="book"
>
  <img
    alt="Book Cover"
    className="book-cover"
    src="https://lightning.od-cdn.com/static/img/no-cover_en_US.a8920a302274ea37cfaecb7cf318890e.jpg"
  />
  <div>
    Loading...
  </div>
</div>
`;
